File: urbanairship-core/src/main/java/com/urbanairship/actions/PromptPermissionAction.java
/* Copyright Airship and Contributors */

package com.urbanairship.actions;

import android.app.Activity;
import android.content.ActivityNotFoundException;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.os.ResultReceiver;
import android.provider.Settings;

import com.urbanairship.UALog;
import com.urbanairship.UAirship;
import com.urbanairship.app.GlobalActivityMonitor;
import com.urbanairship.app.SimpleApplicationListener;
import com.urbanairship.base.Supplier;
import com.urbanairship.json.JsonException;
import com.urbanairship.json.JsonValue;
import com.urbanairship.permission.Permission;
import com.urbanairship.permission.PermissionPromptFallback;
import com.urbanairship.permission.PermissionRequestResult;
import com.urbanairship.permission.PermissionStatus;
import com.urbanairship.permission.PermissionsManager;

import java.util.Objects;
import java.util.concurrent.ExecutionException;

import androidx.annotation.Keep;
import androidx.annotation.MainThread;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

/**
 * An action that prompts for permission.
 * <p>
 * Expected value:
 * - permission: post_notifications, contacts, bluetooth, location, media, mic, or camera
 * - fallback_system_settings: {@code true} to navigate to app settings if the permission is silently denied.
 * - allow_airship_usage: If the permission is granted, any Airship features that depend on the
 * permission will be enabled as well, e.g., enable user notifications on PushManager and push feature
 * on privacy Manager if notifications are allowed.
 * <p>
 * Accepted situations: SITUATION_PUSH_OPENED, SITUATION_WEB_VIEW_INVOCATION,
 * SITUATION_MANUAL_INVOCATION, SITUATION_AUTOMATION, and SITUATION_FOREGROUND_NOTIFICATION_ACTION_BUTTON.
 * <p>
 * Default Result: empty. The actual permission result can be received using a ResultReceiver in the metadata.
 * <p>
 * Default Registration Names: {@link #DEFAULT_REGISTRY_NAME}, {@link #DEFAULT_REGISTRY_SHORT_NAME}
 */
public class PromptPermissionAction extends Action {

    /**
     * Metadata key for a result receiver. Use {@link PermissionResultReceiver} to simplify parsing the result.
     */
    public static final String RECEIVER_METADATA = "com.urbanairship.actions.PromptPermissionActionReceiver";

    /**
     * Default registry name
     */
    @NonNull
    public static final String DEFAULT_REGISTRY_NAME = "prompt_permission_action";

    /**
     * Default registry short name
     */
    @NonNull
    public static final String DEFAULT_REGISTRY_SHORT_NAME = "^pp";

    /**
     * Permission argument key.
     */
    @NonNull
    public static final String PERMISSION_ARG_KEY = "permission";

    /**
     * Enable airship usage argument key.
     */
    @NonNull
    public static final String ENABLE_AIRSHIP_USAGE_ARG_KEY = "enable_airship_usage";

    /**
     * Fallback system settings argument key.
     */
    @NonNull
    public static final String FALLBACK_SYSTEM_SETTINGS_ARG_KEY = "fallback_system_settings";

    /**
     * Permissions result key when using a result receiver.
     */
    @NonNull
    public static final String PERMISSION_RESULT_KEY = "permission";

    /**
     * The starting permission status key when using a result receiver.
     */
    @NonNull
    public static final String BEFORE_PERMISSION_STATUS_RESULT_KEY = "before";

    /**
     * Resulting permission status key when using a result receiver.
     */
    @NonNull
    public static final String AFTER_PERMISSION_STATUS_RESULT_KEY = "after";

    private final Supplier<PermissionsManager> permissionsManagerSupplier;

    public PromptPermissionAction(@NonNull Supplier<PermissionsManager> permissionsManagerSupplier) {
        this.permissionsManagerSupplier = permissionsManagerSupplier;
    }

    @Keep
    public PromptPermissionAction() {
        this(() -> UAirship.shared().getPermissionsManager());
    }

    @Override
    public boolean acceptsArguments(@NonNull ActionArguments arguments) {
        // Validate situation
        switch (arguments.getSituation()) {
            case SITUATION_PUSH_OPENED:
            case SITUATION_WEB_VIEW_INVOCATION:
            case SITUATION_MANUAL_INVOCATION:
            case SITUATION_FOREGROUND_NOTIFICATION_ACTION_BUTTON:
            case SITUATION_AUTOMATION:
                return true;

            case SITUATION_PUSH_RECEIVED:
            case SITUATION_BACKGROUND_NOTIFICATION_ACTION_BUTTON:
            default:
                return false;
        }
    }

    @NonNull
    @Override
    public final ActionResult perform(@NonNull ActionArguments arguments) {
        ResultReceiver resultReceiver = arguments.getMetadata().getParcelable(RECEIVER_METADATA);
        try {
            Args args = parseArg(arguments);
            prompt(args, resultReceiver);
            return ActionResult.newEmptyResult();
        } catch (Exception e) {
            return ActionResult.newErrorResult(e);
        }
    }

    protected Args parseArg(ActionArguments arguments) throws JsonException, IllegalArgumentException {
        return Args.fromJson(arguments.getValue().toJsonValue());
    }

    protected void prompt(@NonNull Args args, @Nullable ResultReceiver resultReceiver) throws ExecutionException, InterruptedException {
        PermissionsManager permissionsManager = Objects.requireNonNull(permissionsManagerSupplier.get());

        permissionsManager.checkPermissionStatus(args.permission, before -> {
            permissionsManager.requestPermission(
                    args.permission,
                    args.enableAirshipUsage,
                    args.fallbackSystemSettings ? PermissionPromptFallback.SystemSettings.INSTANCE : PermissionPromptFallback.None.INSTANCE,
                    requestResult -> sendResult(args.permission, before, requestResult.getPermissionStatus(), resultReceiver)
            );
        });
    }

    public void sendResult(@NonNull Permission permission,
                           @NonNull PermissionStatus before,
                           @NonNull PermissionStatus after,
                           @Nullable ResultReceiver resultReceiver) {

        if (resultReceiver != null) {
            Bundle bundle = new Bundle();
            bundle.putString(PERMISSION_RESULT_KEY, permission.toJsonValue().toString());
            bundle.putString(BEFORE_PERMISSION_STATUS_RESULT_KEY, before.toJsonValue().toString());
            bundle.putString(AFTER_PERMISSION_STATUS_RESULT_KEY, after.toJsonValue().toString());
            resultReceiver.send(Activity.RESULT_OK, bundle);
        }
    }

    @Override
    public boolean shouldRunOnMainThread() {
        return true;
    }

    protected static class Args {

        public final boolean enableAirshipUsage;
        public final boolean fallbackSystemSettings;
        public final Permission permission;

        protected Args(@NonNull Permission permission, boolean enableAirshipUsage, boolean fallbackSystemSettings) {
            this.permission = permission;
            this.enableAirshipUsage = enableAirshipUsage;
            this.fallbackSystemSettings = fallbackSystemSettings;
        }

        @NonNull
        protected static Args fromJson(JsonValue value) throws JsonException {
            Permission permission = Permission.fromJson(value.requireMap().opt(PERMISSION_ARG_KEY));
            boolean enableAirshipUsage = value.requireMap().opt(ENABLE_AIRSHIP_USAGE_ARG_KEY).getBoolean(false);
            boolean fallbackSystemSettings = value.requireMap().opt(FALLBACK_SYSTEM_SETTINGS_ARG_KEY).getBoolean(false);
            return new Args(permission, enableAirshipUsage, fallbackSystemSettings);
        }

    }

}

File: urbanairship-core/src/main/java/com/urbanairship/actions/PromptPermissionAction.kt
/* Copyright Airship and Contributors */
package com.urbanairship.actions

import android.app.Activity
import android.os.ResultReceiver
import androidx.annotation.Keep
import androidx.core.os.bundleOf
import com.urbanairship.Airship
import com.urbanairship.actions.ActionResult.Companion.newEmptyResult
import com.urbanairship.actions.ActionResult.Companion.newErrorResult
import com.urbanairship.actions.PromptPermissionAction.Companion.DEFAULT_NAMES
import com.urbanairship.json.JsonException
import com.urbanairship.json.JsonValue
import com.urbanairship.permission.Permission
import com.urbanairship.permission.PermissionPromptFallback
import com.urbanairship.permission.PermissionStatus
import com.urbanairship.permission.PermissionsManager
import java.util.concurrent.ExecutionException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

/**
 * An action that prompts for permission.
 *
 *
 * Expected value:
 * - permission: post_notifications, contacts, bluetooth, location, media, mic, or camera
 * - fallback_system_settings: `true` to navigate to app settings if the permission is silently denied.
 * - allow_airship_usage: If the permission is granted, any Airship features that depend on the
 * permission will be enabled as well, e.g., enable user notifications on PushManager and push feature
 * on privacy Manager if notifications are allowed.
 *
 *
 * Accepted situations: [Action.Situation.PUSH_OPENED], [Action.Situation.WEB_VIEW_INVOCATION],
 * [Action.Situation.MANUAL_INVOCATION], [Action.Situation.AUTOMATION],
 * and [Action.Situation.FOREGROUND_NOTIFICATION_ACTION_BUTTON].
 *
 *
 * Default Result: empty. The actual permission result can be received using a ResultReceiver in the metadata.
 *
 *
 * Default Registration Name: [DEFAULT_NAMES]
 */
public open class PromptPermissionAction public constructor(
    private val permissionsManagerProvider: () -> PermissionsManager
): Action() {

    private val scope = CoroutineScope(Dispatchers.Main.immediate)

    @Keep
    public constructor() : this({ Airship.permissionsManager })

    override fun acceptsArguments(arguments: ActionArguments): Boolean {
        // Validate situation
        return when (arguments.situation) {
            Situation.PUSH_OPENED,
            Situation.WEB_VIEW_INVOCATION,
            Situation.MANUAL_INVOCATION,
            Situation.FOREGROUND_NOTIFICATION_ACTION_BUTTON,
            Situation.AUTOMATION -> {
                true
            }
            else -> false
        }
    }

    override fun perform(arguments: ActionArguments): ActionResult {
        val resultReceiver = arguments.metadata.getParcelable<ResultReceiver>(RECEIVER_METADATA)

        try {
            val args = parseArg(arguments)
            prompt(args, resultReceiver)
            return newEmptyResult()
        } catch (e: Exception) {
            return newErrorResult(e)
        }
    }

    @Throws(JsonException::class, IllegalArgumentException::class)
    protected open fun parseArg(arguments: ActionArguments): Args {
        return Args.fromJson(arguments.value.toJsonValue())
    }

    @Throws(ExecutionException::class, InterruptedException::class)
    protected fun prompt(args: Args, resultReceiver: ResultReceiver?) {
        val permissionsManager = requireNotNull(permissionsManagerProvider())

        scope.launch {
            val before = permissionsManager.checkPermissionStatus(args.permission)
            val current = permissionsManager.requestPermission(
                permission = args.permission,
                enableAirshipUsageOnGrant = args.enableAirshipUsage,
                fallback = if (args.fallbackSystemSettings) PermissionPromptFallback.SystemSettings else PermissionPromptFallback.None
            )

            sendResult(args.permission, before, current.permissionStatus, resultReceiver)
        }
    }

    public fun sendResult(
        permission: Permission,
        before: PermissionStatus,
        after: PermissionStatus,
        resultReceiver: ResultReceiver?
    ) {
        resultReceiver?.send(
            Activity.RESULT_OK,
            bundleOf(
                PERMISSION_RESULT_KEY to permission.toJsonValue().toString(),
                BEFORE_PERMISSION_STATUS_RESULT_KEY to before.toJsonValue().toString(),
                AFTER_PERMISSION_STATUS_RESULT_KEY to after.toJsonValue().toString()
            )
        )
    }

    override fun shouldRunOnMainThread(): Boolean {
        return true
    }

    public open class Args internal constructor(
        @JvmField public val permission: Permission,
        @JvmField public val enableAirshipUsage: Boolean,
        @JvmField public val fallbackSystemSettings: Boolean
    ) {

        public companion object {

            @Throws(JsonException::class)
            public fun fromJson(value: JsonValue): Args {
                val permission = Permission.fromJson(value.requireMap().opt(PERMISSION_ARG_KEY))

                val enableAirshipUsage = value.requireMap()
                    .opt(ENABLE_AIRSHIP_USAGE_ARG_KEY)
                    .getBoolean(false)

                val fallbackSystemSettings = value.requireMap()
                    .opt(FALLBACK_SYSTEM_SETTINGS_ARG_KEY)
                    .getBoolean(false)

                return Args(permission, enableAirshipUsage, fallbackSystemSettings)
            }
        }
    }

    public companion object {

        /**
         * Metadata key for a result receiver. Use [PermissionResultReceiver] to simplify parsing the result.
         */
        public const val RECEIVER_METADATA: String =
            "com.urbanairship.actions.PromptPermissionActionReceiver"


        /**
         * Default action names.
         */
        public val DEFAULT_NAMES: Set<String> = setOf("prompt_permission_action", "^pp")

        /**
         * Permission argument key.
         */
        public const val PERMISSION_ARG_KEY: String = "permission"

        /**
         * Enable airship usage argument key.
         */
        public const val ENABLE_AIRSHIP_USAGE_ARG_KEY: String = "enable_airship_usage"

        /**
         * Fallback system settings argument key.
         */
        public const val FALLBACK_SYSTEM_SETTINGS_ARG_KEY: String = "fallback_system_settings"

        /**
         * Permissions result key when using a result receiver.
         */
        public const val PERMISSION_RESULT_KEY: String = "permission"

        /**
         * The starting permission status key when using a result receiver.
         */
        public const val BEFORE_PERMISSION_STATUS_RESULT_KEY: String = "before"

        /**
         * Resulting permission status key when using a result receiver.
         */
        public const val AFTER_PERMISSION_STATUS_RESULT_KEY: String = "after"
    }
}
